dataflow/src/main/java/org/checkerframework/dataflow/cfg/CFGBuilder.java

```diff
@@ -649,37 +660,38 @@ public boolean possibleLabels(TypeMirror thrown, Set<Label> labels) {
 
             for (Pair<TypeMirror, Label> pair : catchLabels) {
                 TypeMirror caught = pair.first;
-                boolean canApply = false;
 
-                if (caught instanceof DeclaredType) {
-                    DeclaredType declaredCaught = (DeclaredType) caught;
-                    if (types.isSubtype(declaredThrown, declaredCaught)) {
-                        // No later catch blocks can apply.
-                        labels.add(pair.second);
-                        return true;
-                    } else if (types.isSubtype(declaredCaught, declaredThrown)) {
-                        canApply = true;
-                    }
-                } else {
-                    assert caught instanceof UnionType
-                            : "caught type must be a union or a declared type";
+                if (caught instanceof UnionType) {
                     UnionType caughtUnion = (UnionType) caught;
                     for (TypeMirror alternative : caughtUnion.getAlternatives()) {
                         assert alternative instanceof DeclaredType
                                 : "alternatives of an caught union type must be declared types";
                         DeclaredType declaredAlt = (DeclaredType) alternative;
                         if (types.isSubtype(declaredThrown, declaredAlt)) {
                             // No later catch blocks can apply.
-                            labels.add(pair.second);
+                            Pair<TypeMirror, Label> matchLabel = Pair.of(thrown, pair.second);
+                            labels.add(matchLabel);
                             return true;
                         } else if (types.isSubtype(declaredAlt, declaredThrown)) {
-                            canApply = true;
+                            Pair<TypeMirror, Label> canApplyLabel =
+                                    Pair.of(alternative, pair.second);
+                            labels.add(canApplyLabel);
                         }
                     }
-                }
+                } else {
```

Comment:
Is it really necessary to switch the logic for the then/else branches? That makes it harder to see what changed between them.